### ISSUE TITLE
Unable to use HTTP Action with Microsoft Teams Webhook - Status 400 (Bad Request)

### ISSUE THREAD

**Issue by MaxWinterstein** [drakkan/sftpgo#1266]
I am trying to send a notification to an MS Teams Channel when a customer has provided data. Sadly this always results in an error:

`
 {"level":"debug","time":"2023-04-21T14:57:25.537","sender":"eventmanager","message":"http notification sent, endpoint: https://xxx.webhook.office.com/webhookb2/123@456/IncomingWebhook/789/abc, elapsed: 57.50075ms, status code: 400"}
`
Seetings:

![image](https://user-images.githubusercontent.com/5927148/233671211-841729fa-2def-440a-9529-31b3fcbc1b19.png)

All other addresses work perfectly. 

I even tried to find issues with https://webhook.site:  
![image](https://user-images.githubusercontent.com/5927148/233671862-d0348cb7-8440-4061-a2a5-83562b1556c5.png)

But it seems pretty good. Using their redirect service against my webhook also works well, as well as just old school `curl`. 

I am not sure why M$ is rejecting my request, nor how to debug into that to see what really is set as e.g. headers.

Only thing that stands it is the `@` sign in the address, no idea if this is an issue.

@ drakkan: If needed I can provide you with a valid webhook URL, on the success it replies with `200` and a `1` as data.

**Comment by drakkan** [comment]
Hi,

Perhaps you can try replacing the `@` sign with `%40` and see if that makes any difference.
However I no longer provide free support, sorry. Please check the [support policy](https://github.com/drakkan/sftpgo#support-policy).

**Comment by MaxWinterstein** [comment]
tl;dr;: I think this is a bug 

long story:

I did some digging and found the reason for this:

**The response is missing the `Content-Length` header.**

I ran the requests through [mitmproxy](https://github.com/mitmproxy/mitmproxy) and so I was able to manipulate and replay the requests. 

In case anyone ever is intrested:
- startup `mitmproxyweb` (or console version if you like to suffer)
- modify `sftpgo.json` to listen on port `8082` instead of `8080`
- run with `https_proxy=127.0.0.1:8080 ./sftpgo serve`


**Original request from sftpgo with JSON body, not working**  
![image](https://user-images.githubusercontent.com/5927148/233734172-9688d284-d188-4236-a183-4ba458880273.png)

Response raw:
```
Invalid webhook request - Empty Payload
```

**Original request from sftpgo with no body, not working**  
![image](https://user-images.githubusercontent.com/5927148/233734264-362c68ea-a11d-40f3-845b-e9460d0c2ded.png)

Response raw:
```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML>
<HEAD>
  <TITLE>Length Required</TITLE>
  <META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii">
</HEAD>
<BODY>
  <h2>Length Required</h2>
  <hr>
  <p>HTTP Error 411. The request must be chunked or have a content length.</p>
</BODY>
</HTML>
```

**Modified request, with added `Content-Length` headers, working:**
![image](https://user-images.githubusercontent.com/5927148/233734355-23c72b8d-3499-465a-ad85-e4c450eb543e.png)

As far as I understand RFC9110  https://www.rfc-editor.org/rfc/rfc9110#field.content-length 

>A user agent SHOULD send Content-Length in a request when the method defines a meaning for enclosed content and it is not sending Transfer-Encoding. For example, a user agent normally sends Content-Length in a POST request even when the value is 0 (indicating empty content). A user agent SHOULD NOT send a Content-Length header field when the request message does not contain content and the method semantics do not anticipate such data.

It should be there.

---

As I am no Golang dev, I manually added some debug logs around `internal/common/eventmanager.go` ~ line 1377:
```go
	logger.ErrorToConsole("req %v", req)
	logger.ErrorToConsole("req.header %v", req.Header)
	logger.ErrorToConsole("req.ContentLength %v", req.ContentLength)
```

Which showed me this:

```log
2023-04-21T22:35:49.892 ERR req &{POST https://xxx.webhook.office.com/webhookb2/... HTTP/1.1 1 1 map[Content-Length:[22] Content-Type:[application/json]] {} <nil> 0 [] false m2pfra.webhook.office.com map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc000667aa0}
2023-04-21T22:35:49.892 ERR req.header map[Content-Length:[22] Content-Type:[application/json]]
2023-04-21T22:35:49.892 ERR req.ContentLength 0
```

As said, there is no ContentLength.
 
As far as I understand the usage of [`http.NewRequestWithContext`](https://pkg.go.dev/net/http#NewRequestWithContext) it should provide a `Content-Length`:
> If body is of type *bytes.Buffer, *bytes.Reader, or *strings.Reader, the returned request's ContentLength is set to its exact value (instead of -1), GetBody is populated (so 307 and 308 redirects can replay the body), and Body is set to NoBody if the ContentLength is 0.

The body fed into `NewRequestWithContext` seems to be `io.ReadCloser` and therefore not generating a valid content-length. 

At this point, I am out of time and knowledge about golang.

**Comment by drakkan** [comment]
The content-lenght header is not required in all HTTP versions. This looks like an edge case. I'll leave the issue open but I don't plan to work on this sorry. 

Some nice companies pay for support to get solutions to problems like this and SFTPGo is not a dead project yet thanks to these companies. I am very conflicted and wish I had the test url to work on this but if I fix any reported issue like I did up until recently, SFTPGo will die in the long run because no one has an interest in paying for support and I can't keep spending all this time on SFTPGo for free forever.

If you want to use SFTPGo for free you have no support and you need to be able to fix problems like this yourself sorry. I hope you understand my reasons, thank you

**Comment by drakkan** [comment]
> As an owner of a few things I totally understand it. Time is tight, and pleasing everybody is not possible. TBH, I hope this project gets backed by some foundation or company, and will not die or be forked into a catastrophic tree of unmaintained versions with just hate between the owners. `SFTPGo NG` - hope this will never be a thing... (yeah, I am looking at you, mRemoteNG angry )

I hope this doesn't happen, we'll see, certainly if everyone donated something it wouldn't happen but unfortunately that's not the case

> 
> **It just started working!**
> 
> I left the webhook alone for a while and hoped to find some time to get it fixed myself, or implement something like https://github.com/caronc/apprise on my cluster and let it manage notifications for my stuff. But, out of nowhere, it worked.
> 
> ```json
> {"level":"debug","time":"2023-04-24T14:25:15.545","sender":"eventmanager","message":"http notification sent, endpoint: https://mywebhook, elapsed: 45.096658ms, status code: 400"}
> {"level":"debug","time":"2023-04-25T14:25:19.801","sender":"eventmanager","message":"http notification sent, endpoint: https://mywebhook, elapsed: 4.545993279s, status code: 200"}
> ```
> 
> I did nothing. No restart, no config change, plain nothing. I now just keep my fingers crossed that this keeps working.

this is technically a server-side bug. From what I can understand from your logs SFTPGo connects using HTTP/2 and content length is only required for HTTP/1, maybe they fixed server side

**Comment by drakkan** [comment]
I think the current development version should also add the Content-Length header, as explained it is not required for HTTP/2 since the protocol has the endStream flag to signal the end of the content, so this looks like a server side bug
